### ISSUE TITLE
UTF-8: Implement support in Ruby client library

### ISSUE THREAD

**Issue by ywwg** [prometheus/client_ruby#306]
As in https://github.com/prometheus/client_golang/issues/1369 and https://github.com/prometheus/client_java/issues/916, the Ruby client library needs to be updated to support UTF-8. 

Tasks:

- [ ] Add flag to enable change in validation logic to check that metric and label names are UTF-8 valid instead of the old letters/numbers/underscores/colons set
- [ ] Update exposition format parsers for the new syntax
- [ ] Update PromQL parsers (if any) for the new syntax
- [ ] Update content negotiation logic

For background and references see https://github.com/prometheus/prometheus/issues/13095

**Comment by Sinjo** [comment]
Hey, thanks for the heads up. I've got a rough idea in my head of how those changes will fit into the codebase.

I've given [the proposal](https://github.com/prometheus/proposals/blob/main/proposals/2023-08-21-utf8.md) a read and one thing that stuck out to me is that grouping key labels in the pushgateway client's code aren't mentioned. Are they sticking to the old rules for now?

**Comment by beorn7** [comment]
Very good point. We haven't thought about PGW yet.
That needs more code changes, and some more ideas. **sigh**

**Comment by beorn7** [comment]
For now, I would assume the PGW doesn't support the new UTF-8 names yet.

**Comment by ywwg** [comment]
We are making a note to create a design for this situation

**Comment by Sinjo** [comment]
> That needs more code changes, and some more ideas. **sigh**

Happy to be of service 🙈

**Comment by ywwg** [comment]
@Sinjo are you still interested in working on this? We have done work on the pushgateway so that design is resolved.

**Comment by Sinjo** [comment]
Yeah, I'd like to make some time for it. It doesn't seem like it'll be too hard to bake in at least some experimental support for it.

The biggest potential snag when it comes to our in-process storage of metrics is going to be the way `DirectFileStrore` serialises the metrics to disk to enable multi-process web servers to have all their processes' metrics scraped. I'll need to make sure that code handles UTF-8 label names.

One thing I haven't managed to piece together is how the content negotiation works. Right now we [exclusively serve](https://github.com/prometheus/client_ruby/blob/5514b2be2a417c5fba02479c7ba78096b7eaf4cf/lib/prometheus/client/formats/text.rb) `text/plain; version=0.0.4` format with no conditional behaviours.

Having had a quick glance at https://github.com/prometheus/common/pull/570 it seems that looking for an extra `escaping=allow-utf-8` value in the `Accept` header is enough to enable the new path with unescaped UTF-8 and that it's also possible to pass `underscores`, `dots`, or `values` to that parameter to use one of those escaping schemes. What isn't clear to me is what to do if the registry contains labels with UTF-8 characters, but no `escaping` parameter is passed. Is there a default escaping scheme that's been chosen for compatibility with existing Prometheus servers?

My gut says to stick all of this behind a config flag that makes it clear that all of it is experimental and could change in any minor/patch version. Does that sound right to you?

**Comment by ywwg** [comment]
If no escaping parameter is passed, the metrics producer is expected to apply a default escaping scheme.  In Go, that is determined by the NameEscapingScheme setting in common/model. That is Underscores by default but a user could change it with a configuration value or flag. (I think we haven't implemented this in Go, actually, so I'll file an issue for that :)).  If the metrics producer doesn't escape names, then the metrics will simply be rejected by the metrics consumer as having invalid metric or label names.

UTF-8 will be default-on in Prometheus 3 (and is on by default in the beta) and we should aim for that in all the client libraries. But yes it's acceptable to have this behind a flag for now.

**Comment by Sinjo** [comment]
Cool, that makes sense to me. Thanks!

**Comment by ywwg** [comment]
thank you for contributing! I know only enough ruby to be dangerous so I appreciate the help

**Comment by ywwg** [comment]
@Sinjo any progress?

**Comment by ywwg** [comment]
no worries, unassigning
